File: BOJ 2/16637.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <cstdint>
#include <vector>
#include <string>
#include <iostream>
using namespace std;
typedef int64_t ll;
ll ans = -1e18;
ll calc(char c, ll n1, ll n2) {
    if (c == '*') return n1 * n2;
    else if (c == '+') return n1 + n2;
    else return n1 - n2;
}
void dfs(ll result, int curopidx, vector<char>& op, vector<int>& nums) {
    if (curopidx >= op.size()) {
        ans = max(ans, result);
        return;
    }
    ll res1 = calc(op[curopidx], result, nums[curopidx + 1]);
    dfs(res1, curopidx + 1,op,nums);
    if (curopidx + 1 < op.size()) {
        ll res2 = calc(op[curopidx + 1], nums[curopidx + 1], nums[curopidx + 2]);
        dfs(calc(op[curopidx],result,res2), curopidx + 2, op, nums);
    }
}
int main()
{
    int n;
    string s;
    cin >> n >> s;
    vector<char> op;
    vector<int> nums;
    for (int i = 0; i < s.length(); i++) {
        if (s[i] == '+' || s[i] == '-' || s[i] == '*') op.push_back(s[i]);
        else nums.push_back(s[i] - '0');
    }
    dfs(nums[0], 0, op, nums);
    printf("%d\n", ans);
    return 0;
}

File: BOJ(Java 공부용)/16637.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Main {

    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 1, -1, 0};
    static int[] goy = {1, 0, 0, -1};
    static long ans = (long) -1e18;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String s = br.readLine();
        List<Character> op = new ArrayList<>();
        List<Integer> nums = new ArrayList<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '+' || s.charAt(i) == '-' || s.charAt(i) == '*') op.add(s.charAt(i));
            else nums.add(s.charAt(i) - '0');
        }
        dfs(nums.get(0), 0, op, nums);
        bw.write(ans + "");
        bw.flush();
        bw.close();
    }

    private static void dfs(long result, int curopidx, List<Character> op, List<Integer> nums) {
        if (curopidx >= op.size()) {
            ans = Math.max(ans, result);
            return;
        }
        long res1 = calc(op.get(curopidx), result, nums.get(curopidx + 1));
        dfs(res1, curopidx + 1, op, nums);
        if (curopidx + 1 < op.size()) {
            long res2 = calc(op.get(curopidx + 1), nums.get(curopidx + 1), nums.get(curopidx + 2));
            dfs(calc(op.get(curopidx), result, res2), curopidx + 2, op, nums);
        }
    }

    private static long calc(char c, long n1, long n2) {
        if (c == '+') return n1 + n2;
        else if (c == '-') return n1 - n2;
        else return n1 * n2;
    }
}